### PR TITLE
zehra_week3

### PR DESCRIPTION
**Pull request by zehrakopar**
JavaScript 2 week 3 homework

### CODE REVIEW

File: Week3/homework/step2-1.js [hunk 1]
```diff
@@ -1,9 +1,7 @@
 'use strict';
 
 function foo(func) {
-  // What to do here?
-  // Replace this comment and the next line with your code
-  console.log(func);
+  return func();
```
Comment (remarcmij) [inline Week3/homework/step2-1.js]: You may, but do not have to return the return value of `func` as the return value of `foo`. In this case that return value will be `undefined` as `bar` does not return a value:

```js
function foo(func) {
  func();
}
```

File: Week3/homework/step2-2.js [hunk 2]
```diff
@@ -3,18 +3,24 @@
 function threeFive(startIndex, stopIndex, threeCallback, fiveCallback) {
   const numbers = [];
 
-  // Replace this comment and the next line with your code
-  console.log(startIndex, stopIndex, threeCallback, fiveCallback, numbers);
+  for (let i = startIndex; i <= stopIndex; i++) {
```
Comment (remarcmij) [inline Week3/homework/step2-2.js]: The assignment called for a two step approach: first create an array and then iterate through the array:

```js
for (let i = startIndex; i <= stopIndex; i++) {
    numbers.push(i);
}

for (let i = 0; i < numbers.length; i++ {
  ...
}
```

File: Week3/homework/step2-2.js [hunk 3]
```diff
@@ -3,18 +3,24 @@
...
+    if (i % 5 === 0) {
+      fiveCallback(i);
+    }
+  }
 }
 
 function sayThree(number) {
-  // Replace this comment and the next line with your code
-  console.log(number);
+  return number;
```
Comment (remarcmij) [inline Week3/homework/step2-2.js]: These functions were supposed to 'say' something, e.g.:

```
The number 12 is divisible by 3. They don't need to return anything.
```

File: Week3/homework/step2-3.js [hunk 4]
```diff
@@ -18,22 +19,28 @@ function repeatStringNumTimesWithWhile(str, num) {
...
-
 // Use a 'do...while' loop
 function repeatStringNumTimesWithDoWhile(str, num) {
   // eslint-disable-next-line prefer-const
   let result = '';
 
-  // Replace this comment and the next line with your code
-  console.log(str, num, result);
-
+  if (num > 0) {
```
Comment (remarcmij) [inline Week3/homework/step2-3.js]: 👍 for checking the value of `num` before entering the loop and skipping the look altogether is `num <= 0`.

File: Week3/homework/step2-4.js [hunk 5]
```diff
@@ -1,7 +1,9 @@
 'use strict';
 
 function Dog() {
-  // add your code here
+  this.name = 'Albert';
+  this.color = 'white';
+  this.numLegs = 4;
 }
 
 const hound = new Dog();
```
Comment (remarcmij) [inline Week3/homework/step2-4.js]: 👍

File: Week3/homework/step2-5.js [hunk 6]
```diff
@@ -4,8 +4,11 @@ function multiplyAll(arr) {
...
-  // Replace this comment and the next line with your code
-  console.log(arr, product);
+  for (let i = 0; i < arr.length; i++) {
+    for (let j = 0; j < arr[i].length; j++) {
+      product *= arr[i][j];
+    }
+  }
 
   return product;
 }
```
Comment (remarcmij) [inline Week3/homework/step2-5.js]: 👍

File: Week3/homework/step2-6.js [hunk 7]
```diff
@@ -4,13 +4,27 @@ const arr2d = [[1, 2], [3, 4], [5, 6]];
...
+    for (let j = 0; j < arr[i].length; j++) {
+      for (let k = 0; k < arr[i][j].length; k++) {
+        flattenedArr.push(arr[i][j][k]);
+      }
+    }
+  }
+  return flattenedArr;
 }
 
 console.log(flattenArray2d(arr2d)); // -> [1, 2, 3, 4, 5, 6]
```
Comment (remarcmij) [inline Week3/homework/step2-6.js]: Okay, this works fine for the arrays provided. Can you think of a way to deal with arrays of any dimension?

File: Week3/homework/step2-7.js [hunk 8]
```diff
@@ -20,4 +20,11 @@ f2(y);
...
 
-// Add your explanation as a comment here
+// in the f1 function we use (const x = 9); The variable equals to a primitive value(undefined, null, boolean, string, and numbers). In Pass by Value,
+// Function is called by directly passing the value of the variable as the argument. Changing the argument inside the function
+// doesn’t affect the variable passed from outside the function. So changing the value of the variable with a function never
+// changes the underlying primitive (String or number).
+
+// in the f2 function we use (const y = { x: 9 }); it equals to an object. In Pass by Reference, Function is called by directly
+// passing the reference/address of the variable as the argument. Changing the argument inside the function affect the variable
+// passed from outside the function. The reference/address of the objects and arrays same but the values at that addresses changed.
```
Comment (remarcmij) [inline Week3/homework/step2-7.js]: I'm happy with your explanation.

File: Week3/homework/step3-bonus.js [hunk 9]
```diff
@@ -1,14 +1,26 @@
...
+console.log(makeUnique(myNumbers));
+
+function unique(arr) {
+  return [...new Set(arr)];
+}
+console.log(unique(values));
+console.log(unique(myColors));
+console.log(unique(myNumbers));
+
+// I ve learned the both methods from stack overflow
```
Comment (remarcmij) [inline Week3/homework/step3-bonus.js]: Both solutions are fine. It's okay to use stack overflow but make sure you can explain how the code works.

File: Week3/homework/step3.js [hunk 10]
```diff
@@ -2,7 +2,10 @@
...
 function createBase(base) {
   // Replace this comment and the next line with your code
-  console.log(base);
+  function add(number) {
+    return base + number;
+  }
+  return add;
 }
 
 const addSix = createBase(6);
```
Comment (remarcmij) [inline Week3/homework/step3.js]: You can also directly return an anonymous function, e.g.:

```js
function createBase(base) {
  return function(number) {
    return base + number;
  };
}
```

Or, if you are adventurous, use arrow functions:

```js
const createBase = base => number => base + number;
```

File: Week3/homework/step2-1.js (hunk 1 above)
Comment (zehrakopar) [reply Week3/homework/step2-1.js]: Got it.

File: Week3/homework/step2-2.js (hunk 2 above)
Comment (zehrakopar) [reply Week3/homework/step2-2.js]: I ve changed....


'use strict';

function threeFive(startIndex, stopIndex, threeCallback, fiveCallback) {
  const numbers = [];

  for (let i = startIndex; i <= stopIndex; i++) {
    numbers.push(i);
  }
  for (let i = 0; i < numbers.length; i++) {
    const number = numbers[i];
    if (number % 3 === 0) {
      threeCallback(number);
    }
    if (number % 5 === 0) {
      fiveCallback(number);
    }
  }
};

  function sayThree(number) {
  return `The number ${number} is divisible by 3.`;
};

function sayFive(number) {
  return `The number ${number} is divisible by 5.`;
};

threeFive(10, 15, sayThree, sayFive);

// Do not change or remove anything below this line
module.exports = threeFive;

Comment (zehrakopar) [reply Week3/homework/step2-2.js]: i ve changed.

**Pull request closed without merging**